src/v/s3/signature.cc

```diff
@@ -0,0 +1,392 @@
+    vassert(ret > 0, "Invalid date format string");
+    return ss::sstring(out_str.data(), ret);
+}
+
+ss::sstring time_source::format_date() const { return format("%Y%m%d"); }
+
+ss::sstring time_source::format_datetime() const {
+    return format("%Y%m%dT%H%M%SZ");
+}
+
+timestamp time_source::default_source() {
+    return std::chrono::system_clock::now();
+}
+
+// signature_v4 //
+
+enum {
+    sha256_digest_length = 32,
+};
+
+using hmac_digest = std::array<char, sha256_digest_length>;
+
+template<class Container>
+static ss::sstring hexdigest(Container const& digest) {
+    std::array<uint8_t, sha256_digest_length> result{};
+    std::memcpy(result.data(), digest.data(), digest.size());
+    return to_hex(bytes_view{result.data(), result.size()});
+}
+
+static hmac_digest hmac(std::string_view key, std::string_view value) {
+    hmac_digest digest{};
+    int ec = gnutls_hmac_fast(
+      GNUTLS_MAC_SHA256,
+      key.data(),
+      key.size(),
+      value.data(),
+      value.size(),
+      digest.data());
+    vassert(ec >= 0, "HMAC error '{}'", ec);
+    return digest;
+}
+
+static result<ss::sstring> sha_256(std::string_view str) {
+    std::array<uint8_t, sha256_digest_length> hash{};
+    int ec = gnutls_hash_fast(
+      GNUTLS_DIG_SHA256,
+      str.data(),
+      str.size(),
+      static_cast<void*>(hash.data()));
+    if (ec < 0) {
+        vlog(s3_log.error, "sha256 error '{}'", ec);
+        return make_error_code(ec);
+    }
+    return to_hex(bytes_view{hash.data(), hash.size()});
+}
+
+result<ss::sstring> signature_v4::gen_sig_key(
+  std::string_view key,
+  std::string_view datestr,
+  std::string_view region,
+  std::string_view service) {
+    hmac_digest digest{};
+    ss::sstring initial_key = "AWS4";
+    initial_key.append(key.data(), key.size());
+    digest = hmac(std::string_view(initial_key), datestr);
+    digest = hmac(std::string_view(digest.data(), digest.size()), region);
+    digest = hmac(std::string_view(digest.data(), digest.size()), service);
+    digest = hmac(
+      std::string_view(digest.data(), digest.size()), "aws4_request");
+    return ss::sstring{digest.data(), digest.size()};
+}
+
+static void tolower(ss::sstring& str) {
+    for (auto& c : str) {
+        c = std::tolower(c);
+    }
+}
+
+static void append_hex_utf8(ss::sstring& result, char ch) {
+    bytes b = {static_cast<uint8_t>(ch)};
+    result.append("%", 1);
+    auto h = to_hex(b);
+    result.append(h.data(), h.size());
+}
+
+static ss::sstring uri_encode(ss::sstring const& input, bool encode_slash) {
+    // The function defined here:
+    //     https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-header-based-auth.html
+    ss::sstring result;
+    for (auto ch : input) {
+        if (
+          (ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z')
+          || (ch >= '0' && ch <= '9') || (ch == '_') || (ch == '-')
+          || (ch == '~') || (ch == '.')) {
```

Comment:
nit: we also have compile time regular expression library in tree. here is an example usage of something fairly similar: https://github.com/vectorizedio/redpanda/blob/dev/src/v/kafka/security/scram_algorithm.cc#L75

Reply:
A note/comment about it would be good.